spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala

```diff
@@ -152,3 +153,17 @@ case class CometBatchScanExec(wrapped: BatchScanExec, runtimeFilters: Seq[Expres
 
   override def supportsColumnar: Boolean = true
 }
+
+object CometBatchScanExec {
+
+  def isSchemaSupported(schema: StructType): Boolean =
```

Comment:
What was the reason of copying `isSchemaSupported` and `isTypeSupported` here?
It looks they are still used only in `CometSparkSessionExtensions.scala`, so that we can keep using the original one?
Also there is another copy in `CometScanExec.scala`

Reply:
Currently it's very hard coded that everything in Comet supports the exact same data types. This was an attempt to start breaking out supported data types for different functionality within Comet. Just because I want to support struct types in expressions, doesn't mean I have to support it in the Scan operations. And when the Scans are updated to support each complex type, they can then opt-in to supporting them. And CometBatchScanExec and CometScanExec don't need to be updated at the same time to support the same types either.

We could get away with just delegating back to the CometSparkSessionExtensions check with the new allowComplex flag, but it seems to make a lot more sense to let each operator opt-in to exactly what it supports.

Reply:
Got it. Let's delete the original one then? With this PR it is no longer used I think.

Also this object should be package private unless we add a test for it. `isTypeSupported` should be `private` as well

Reply:
It's still used just by `shouldApplyRowToColumnar`, but to go with this theme I could move that onto the CometRowToColumnarExec object

Reply:
I don't think I can make `object CometBatchScanExec` private? Since it's being accessed from `org.apache.comet` and it's in `org.apache.spark.sql.comet`. I think all I can do is make `isTypeSupported` private?

Reply:
I see. Is it possible to create a base trait? So that we can share the logic for be basic types between different objects.
Then perhaps new `CometRowToColumnarExec` object can inherit it and call super + extra type match.

I cannot think cases like a type supported only in `CometBatchScanExec` but not in `CometScanExec`...

Reply:
I'll try to make a trait to share at least all of the common types supported "everywhere"

Reply:
Naming is hard, but something like
```scala
trait DataTypeSupport {
  def isAdditionallySupported(dt: DataType): Boolean = false

  private def isGloballySupported(dt: DataType): Boolean = dt match {
    case BooleanType | ByteType | ShortType | IntegerType | LongType | FloatType | DoubleType |
        BinaryType | StringType | _: DecimalType | DateType | TimestampType =>
      true
    case t: DataType if t.typeName == "timestamp_ntz" => true
    case _ => false
  }

  def isSchemaSupported(struct: StructType): Boolean = {
    struct.fields.map(_.dataType).forall(isTypeSupported)
  }

  def isTypeSupported(dt: DataType): Boolean = {
    if (isGloballySupported(dt) || isAdditionallySupported(dt)) {
      // If complex types are supported, we additionally want to recurse into their children
      dt match {
        case StructType(fields) => fields.map(_.dataType).forall(isTypeSupported)
        case ArrayType(elementType, _) => isTypeSupported(elementType)
        case MapType(keyType, valueType, _) =>
          isTypeSupported(keyType) && isTypeSupported(valueType)
        // Not a complex type
        case _ => true
      }
    } else {
      false
    }
  }
}
```
?

Reply:
Pushed, let me know how it looks. Not a huge fan of the function names but couldn't come up with anything better. It could maybe be expanded to other operators in the future if they need to opt in or out of certain data types support.